### DISCUSSION TITLE
Should Jotai support atom graph cycles?

### DISCUSSION THREAD

**Discussion by dmaskasky** [pmndrs/jotai#2823] [Q&A] [answered]
```ts
const a = atom((get) => {
  return get(b)
})
const b = atom((get) => {
  return get(a)
})
```

A more real world example would be:
```ts
const valueAtom = atom(0)
const effectAtom = atomEffect((get) => {
  console.log('derived is now', get(derivedAtom))
})
const derivedAtom = atom((get) => {
  get(effectAtom) // mount effect
  return get(valueAtom)
})
```

**Comment by dai-shi** [comment, marked as answer]
I don't think we should support it. It's mathematically wrong, or undefined, right?
I think the cycles would cause infinite loops, or should error with cycle / infinite loop detection.

I think it's reasonable for `atomEffect` to have its own workaround as effect can be imperative instead of declarative.

**Reply by dmaskasky** [reply]
Thanks for clarifying this requirement.

**Reply by na2hiro** [reply]
@dai-shi what's the correct way to ensure that an effect is hooked to an atom always? Before https://github.com/pmndrs/jotai/pull/2742, I used to make such circular dependency.

**Reply by dai-shi** [reply]
use jotai-effect? or do I misunderstand your question?

**Reply by dmaskasky** [reply]
You can use the `withAtomEffect` api.

```ts
const countAtom = atom(0)
export const countWithEffect = withAtomEffect(countAtom, (get, set) => {
  console.log('count is now', get(countAtom))
})
```

Prior to the intro of `withAtomEffect`, you would need to make a tree structure (no cycles).
```ts
const countAtom = atom(0)
const effectAtom = atom((get, set) => {
  console.log('count is now', get(countAtom))
})
export const countWithEffect = atom(
  (get) => get(countAtom),
  (get, set, valueOrFn: SetStateAction<number>) => set(countAtom, valueOrFn)
)
```

**Reply by na2hiro** [reply]
@dmaskasky Thanks, I got it now!

I had migrated from Recoil and I was a bit confused that it's not by default hooked into observing atoms (With [Recoil's atom effect](https://recoiljs.org/docs/guides/atom-effects/), `effects` array is specified in its atom definition).

It might be useful to add this example in [`withAtomEffect` section](https://jotai.org/docs/extensions/effect#withatomeffect), as the example there inlines `atom(0)` and it's not obvious how to achieve this pattern (getting `valuesAtom` makes a circular dependency).

**Reply by dmaskasky** [reply]
> getting `valuesAtom` makes a circular dependency

This is not actually the case because atomEffect [does not cause infinite loops](https://jotai.org/docs/extensions/effect#:~:text=Resistant%20To%20Infinite%20Loops) by default.

For instance, the following test passes.
```ts
it('does not cause infinite loops when it references itself', async () => {
  const countWithEffectAtom = withAtomEffect(atom(0), (get, set) => {
    get(countWithEffectAtom)
    set(countWithEffectAtom, (v) => v + 1)
  })
  const store = createStore()
  store.sub(countWithEffectAtom, () => {})
  await Promise.resolve()
  expect(store.get(countWithEffectAtom)).toBe(1)
  store.set(countWithEffectAtom, (v) => ++v)
  await Promise.resolve()
  expect(store.get(countWithEffectAtom)).toBe(3)
})
```

**Reply by na2hiro** [reply]
I tested with `withAtomEffect` in this sandbox https://stackblitz.com/edit/react-ts-m9mskdnu?file=index.tsx

* Atom effects are not set up when used with `useHydrateAtoms()`. Isn't it a bug?
* If set up like as you provided initially (See 2, WithoutHydration), the effect runs initially but doesn't run for the succeeding updates. If I update the code to `get()` is passed the return value of `withAtomEffect`, it works as expected (See 3, WithoutHydration). It seems that the first arg of `withAtomEffect()` should be always inlined to avoid this pitfall. I wonder if this sounds true

**Reply by dmaskasky** [reply]
Confirmed. Reproduction: https://stackblitz.com/edit/react-ts-euct7mig?file=index.tsx

I added these two issues:
 - https://github.com/jotaijs/jotai-effect/issues/46
 - https://github.com/jotaijs/jotai-effect/issues/47

**Reply by na2hiro** [reply]
@dmaskasky Thanks!

**Reply by dmaskasky** [reply]
@na2hiro Both issues are now resolved. Thank you for bringing this to my attention!
Please upgrade to [v1.0.6](https://www.npmjs.com/package/jotai-effect/v/1.0.6). [Release notes](https://github.com/jotaijs/jotai-effect/releases/tag/v1.0.6)

**Reply by na2hiro** [reply]
Great, I could hook up the effects the way I wanted now. Thanks a lot for the quick resolution!
